README.md
# README

### Prerequisite 1: Haskell Stack
In Linux:
```
$ wget -qO- https://get.haskellstack.org/ | sh
```
In Mac:
```
$ brew install haskell-stack
```
See https://docs/haskellstack.org/en/stable/README/#how-to-install for details.

## Prerequisite2: 
In Linux:
```
$ sudo apt-get install nkf gnuplot
```

## Install
```
$ cd <dynet-tools>
$ stack build
```

## Contact ##

* Repo owner: [Daisuke Bekki](http://www.is.ocha.ac.jp/~bekki/)
* [Learn Markdown](https://bitbucket.org/tutorials/markdowndemo)



ChangeLog.md
# Changelog for dytest

## Unreleased changes


test/Test.hs
{-# OPTIONS -Wall #-}
{-# LANGUAGE OverloadedStrings #-}

import qualified System.Environment as E --base
import qualified DyNet.Core as D     --dynet-haskell

main :: IO ()
main = do
  argv <- E.getArgs
  _ <- D.initialize' argv
  model <- D.createModel
  _ <- D.addParameters' model [10::Int, 10::Int]
  putStrLn "ok"
  


app/drel/DRel.hs
{-# LANGUAGE OverloadedStrings #-}

module DRel where

import Control.Monad ( forM, forM_, liftM )
import qualified System.Environment as E -- base
import qualified Data.Text as T          -- text
import qualified Data.Text.IO as T       -- text
import qualified DNN.Utils as U          -- keras-dnn
import qualified DyNet.Core as Core      -- DyNet
import qualified DyNet.Expr as Expr      -- DyNet
import qualified DyNet.Train as Train    -- DyNet
--import qualified DyNet.IO as IO          -- DyNet
import qualified DyNet.Vector as Vec     -- DyNet

-- | prepares a float vector of dimension dim
zeroVector :: Int -> IO (Vec.Vector Float)
zeroVector dim = Vec.fromList (take dim $ repeat $ (0::Float))

-- | converts Data.Text to Float
text2float :: T.Text -> Float
text2float text = (read $ T.unpack text)::Float

-- | sets a value (a list of Data.Texst) of a (pointer to a) vector vec
setValue :: (Vec.Vector Float) -> [T.Text] -> IO()
setValue vec vals = mapM_ (\(val,ind) -> Vec.insert vec ind $ text2float val) $ zip vals [0..]

-- | converts a list of floats so that a float greater than a given threshold is True, and False otherwise.
--screen :: Float -> [Float] -> [Bool]
--screen threshold = map (>= threshold)

main :: IO()
main = do
  let hidden_dim1 = 100
      epochs = 100::Int
      --threshold = 0.55::Float
  argv <- do
          a <- E.getArgs
          Core.initialize' a
  model <- Core.createModel
  trainer <- Train.createSimpleSGDTrainer model 0.1 0.0

  datadirroot <- U.getPath
  let datadir = datadirroot ++ "app/DRel/data/"
      
  feat_train    <- U.openCSV "," $ datadir ++ "feat_train.txt"
  feat_test     <- U.openCSV "," $ datadir ++ "feat_test.txt"
  words_train   <- U.openCSV "," $ datadir ++ "words_train.txt"
  words_test    <- U.openCSV "," $ datadir ++ "words_test.txt"
  predrel_train <- U.openCSV "," $ datadir ++ "predrel_train.txt"
  predrel_test  <- U.openCSV "," $ datadir ++ "predrel_test.txt"
  drel_train    <- U.openCSV "," $ datadir ++ "drel_train.txt"
  drel_test     <- U.openCSV "," $ datadir ++ "drel_test.txt"
  drel_labels   <- liftM T.lines $ T.readFile $ datadir ++ "drel.txt.sorted"
  
  let feat_dim = length $ head feat_train
      words_dim = length $ head words_train
      predrel_dim = length $ head predrel_train
      drel_dim = length $ head drel_train

  -- print_ =<< getDim h
  
  p_W1 <- Core.addParameters' model [hidden_dim1, feat_dim+predrel_dim]
  p_b1 <- Core.addParameters' model [hidden_dim1]
  p_W2 <- Core.addParameters' model [drel_dim, hidden_dim1]
  p_b2 <- Core.addParameters' model [drel_dim]
  p_emb <- Core.addLookupParameters' model 0 (256::Vec.Int64)

  Core.withNewComputationGraph $ \cg -> do
    -- | 入力層
    fVals <- zeroVector feat_dim
    f <- Expr.input cg [feat_dim] fVals
    pVals <- zeroVector predrel_dim
    p <- Expr.input cg [predrel_dim] pVals
    x <- Expr.concat' [f,p]
    Core.print_ =<< Core.getDim f
    Core.print_ =<< Core.getDim p
    Core.print_ =<< Core.getDim x
    -- | 出力層
    yVals <- zeroVector feat_dim
    y <- Expr.input cg [drel_dim] yVals
    -- | １層目
    _W1 <- Expr.parameter cg p_W1
    b1 <- Expr.parameter cg p_b1
    h1 <- Expr.selu $ _W1 `Expr.mul` x `Expr.add` b1
    -- | ２層目
    _W2 <- Expr.parameter cg p_W2
    -- b2 <- Expr.parameter cg p_b2
    y_hat <- Expr.logistic $ _W2 `Expr.mul` h1 -- `Expr.add` b2
    lossExp <- Expr.squaredDistance y_hat y
    -- | Training
    _ <- forM [1..epochs] $ \epoch -> do
           loss' <- forM (zip3 feat_train predrel_train drel_train) $ \(feat,predrel,drel) -> do
             setValue fVals feat
             setValue pVals predrel
             setValue yVals drel
             loss <- Core.asScalar =<< Core.forward cg lossExp
             Core.backward cg lossExp
             Train.update trainer 1.0
             return loss
           putStrLn $ "Epoch " ++ show epoch
                      ++ ": Average loss = " ++ show ((sum loss') / (realToFrac $ length loss'))
    -- | Test
    y_hatT <- forM (zip feat_test predrel_test) $ \(feat,predrel) -> do
                setValue fVals feat
                setValue pVals predrel
                Vec.toList =<< Core.asVector =<< Core.forward cg y_hat
    forM_ [0.1, 0.2, 0.28, 0.29, 0.3, 0.31, 0.32, 0.33, 0.4, 0.5, 0.52, 0.55, 0.57, 0.6, 0.7, 0.8, 0.9] $ \threshold -> do 
      result <- U.initializeMultilabelResult
                   drel_labels
                   (map (map (>= threshold)) y_hatT)
                   (map (map ((>= 1.0) . text2float)) drel_test)
    --print $ U.classificationResult result
      putStrLn $ "Threshold: " ++ (show threshold)
               ++ ", Accuracy: " ++ (show $ U.accuracy result)
               ++ ", Precision: " ++ (show $ U.precision result)
               ++ ", Recall: " ++ (show $ U.recall result)
               ++ ", F1: " ++ (show $ U.f1 result)




app/jsem/jsemClassify.hs
{-# LANGUAGE OverloadedStrings, DeriveGeneric #-}

--module JSeMclassify where

import System.FilePath ((</>),(<.>))
--import Control.Exception as E
import qualified Data.Text as T    --text
import qualified Data.Text.IO as T --text
import qualified Data.Aeson as A   --aeson

import qualified DyNet.Core as D      --dynet-haskell
import DyNet.Tools.Juman as Juman     --dynet-tools
import DyNet.Simple as D              --dynet-tools
import DyNet.Simple.LSTMclassifier    --dynet-tools

-- | JSeM classifier
--   Usage : stack exec jsem-run 1-128-10-256-412-88-20-100 test.txt
--               1-128-10-256-412-88-20-100: hyperparameters
--               test.txt: a text file of an inference data (the last line of which is the hypothesis)
main :: IO()
main = do
  let homedir = "/home/bekki/program/dynet-tools"
      loaddir = homedir </> "app" </> "jsem" </> "jsem_models"
--  E.catch (do
  -- | 初期化。第一引数はjson形式ファイル、第二引数は前提文・帰結文が書かれたテキストファイル。
  (aesonFileName:premisehypoFileName:_) <- D.setup
  -- | jsonファルからハイパーパラメータを読み出す。
  maybeHyperParams <- (A.decodeFileStrict $ loaddir </> aesonFileName <.> "json")
  hyperParams <- case maybeHyperParams of
        Nothing -> ioError $ userError $ aesonFileName ++ " not well-formed."
        Just json -> return json
  -- | テキストファイルから
  (premiseTx,hypoTx) <- (\lns -> (T.concat (init lns), last lns)) <$> T.lines <$> T.readFile premisehypoFileName
  premise <- Juman.processTextByJumanOnline premiseTx
  hypo <- Juman.processTextByJumanOnline hypoTx
  model <- D.createModel
  params <- setLSTMparams model hyperParams
  D.loadModel model $ loaddir </> (showHyperParams hyperParams) <.> "model"
  D.classify (buildClassifier params) (labels params) (premise,hypo) >>= T.putStrLn
--    ) $ \(E.SomeException e) -> print e



src/DyNet/Tools/Chart.hs
{-# OPTIONS -Wall #-}
{-# LANGUAGE OverloadedStrings #-}

module DyNet.Tools.Chart (
  drawLearningCurve
  ) where

import Data.List (transpose)
import Graphics.Gnuplot.Simple
import DyNet.Simple (LearningChart)

-- | LearningChartは各行が
-- (0, [107.26212, 112.046524 94.8856 91.56159 85.57186 25.32702])
-- (1, [58.097065, 114.710655 101.360596 98.431046 84.57611 24.729351])
-- なので、これから
-- (sty "batch-1", [(0,107.26212),(1,58.09765),...])
-- を剥ぎとって（どういう意味？？
-- (0, [112.046524 94.8856 91.56159 85.57186 25.32702])
-- (1, [114.710655 101.360596 98.431046 84.57611 24.729351])
-- を返して再帰する。

testData :: LearningChart
testData = [
  (0, [107.26212, 112.046524, 94.8856,    91.56159,  85.57186,  25.32702]),
  (1, [58.097065, 114.710655, 101.360596, 98.431046, 84.57611,  24.729351]),
  (2, [61.99805,  89.96048,   86.6363,    89.605865, 88.304985, 23.252378]),
  (3, [62.932064, 91.324265,  83.364265,  85.87307,  82.94884,  20.744677]),
  (4, [48.18716,  97.32671,   84.417404,  85.4241,   79.75656,  17.75763]),
  (5, [46.595695, 83.82408,   73.23103,   79.69995,  80.71408,  15.31874]),
  (6, [47.275078, 79.2172,    67.13917,   73.52796,  74.99438,  11.9399395]),
  (7, [38.305492, 82.1474,    64.91645,   70.82271,  74.180565, 8.756168]),
  (8, [32.533005, 73.59252,   51.76041,   63.517025, 76.30533,  6.813904]),
  (9, [29.234098, 68.950035,  43.700676,  59.53903,  70.3834,   4.861396]),
  (10,[21.261477, 75.850494,  42.84322,   58.941006, 65.03832,  3.999124])
  ]

drawLearningCurve :: FilePath -> String -> LearningChart -> IO()
drawLearningCurve filepath title learningchart = do
  let maxepoch  = fromIntegral $ length learningchart
      dat       = transpose $ map (\(epoch,losses) -> map (\loss -> (fromIntegral epoch, (realToFrac loss)::Double)) losses) learningchart
      styleddat = zip (map (\epoch -> let linetitle = "batch-" ++ (show epoch)
                                      in PlotStyle LinesPoints (CustomStyle [LineTitle linetitle])
                                      ) [(0::Int)..]) dat -- | [(PlotStyle, [(Double,Double)])]
      graphstyle = [(PNG filepath),(Title title),(XLabel "epoch"),(YLabel "loss"),(XRange (0,maxepoch))]
  plotPathsStyle graphstyle styleddat



src/DyNet/Tools/Juman.hs
{-# OPTIONS -Wall #-}
{-# LANGUAGE OverloadedStrings, ExtendedDefaultRules #-}
{-# OPTIONS_GHC -fno-warn-type-defaults #-}

module DyNet.Tools.Juman (
  JumanEntry (..),
  Kihon,
  Hinsi,
  processTextsByJuman,
  processTextByJumanOnline,
  buildDictionary,
  cutLessFreqWords
  --juman2tensor,
  --unzipJumanTensorWithMask,
  --jumanTensor2BagOfWords,
  --printJumanTensorParams,
  ) where

--import System.FilePath ((</>)) --filepath
import Control.Monad (guard,forM)  --base
--import qualified System.IO as IO   --base
import qualified Data.List as L    --base
import qualified Shelly as S       --shelly
import qualified Data.Text as T    --text
--import qualified Data.Text.IO as T --text
import qualified DyNet.Tools.Utils as U  --dynet-tools
default (T.Text)

data JumanEntry = JumanEntry {
  hyoso :: T.Text,
  yomi :: T.Text,
  kihon :: T.Text,
  pos :: T.Text,
  subpos :: T.Text
  } | EOS | Error T.Text deriving (Eq, Show)

-- | takes a text, process it with JUMAN, and returns the result.  Used internally.
callJuman :: T.Text -> S.Sh T.Text
callJuman text = S.silently $ S.escaping False $ S.cmd $ S.fromText $ T.concat ["echo ", purifyText text, " | juman"]
  where purifyText tx = -- | removes occurrences of non-letters from an input text.  
          case T.uncons tx of -- remove a non-literal symbol at the beginning of a sentence (if any)
            Nothing -> T.empty
            Just (c,t) | T.any (==c) "("          -> T.cons '（' $ purifyText t
                       | T.any (==c) ")"          -> T.cons '）' $ purifyText t
                       | T.any (==c) ","          -> T.cons '，' $ purifyText t
                       | T.any (==c) "#"          -> T.cons '＃' $ purifyText t
                       | T.any (==c) "`'^><_&;\\" -> T.cons '*'  $ purifyText t  -- ignore meaningless symbols
                       | otherwise                -> T.cons c    $ purifyText t

callJuman' :: T.Text -> S.Sh [JumanEntry]
callJuman' text = do
  jumanlines <- T.lines <$> callJuman text
  return $ map jumanOutput2Entry $ L.filter (\t -> not $ T.isPrefixOf "@" t) jumanlines

type Kihon = T.Text
type Hinsi = T.Text

-- | takes a filepath and a list of texts,
--   process each text with `callJuman',
--   and write the result in a specified fiie (unless it already exists)
processTextsByJuman :: FilePath -> [T.Text] -> IO [[(Kihon, Hinsi)]]
processTextsByJuman filepathname texts = S.shelly $ do
  let filepath = S.fromText $ T.pack filepathname
      for = flip map
  exist <- S.test_f filepath
  if exist -- if the file already exists 
     then do
          S.echo_n_err $ T.pack $ filepathname ++ " found. To re-run juman, delete " ++ filepathname ++ " and run the program again.\n"
          -- "k1#p1,...,kn#pn \n k1#p1,...,kn#pn \n ..." >>= ["k1#p1,...,kn#pn", "k1#p1,...,kn#pn", ...]
          lins <- T.lines <$> S.readfile filepath
          return $ for lins $ \line -> do
                                pairs <- T.split (==',') line
                                let (k:p:_) = T.split (=='#') pairs
                                return (k,p)
     else do
          S.echo_n_err $ T.pack $ show (length texts) ++ " texts processed.\n"
          forM texts $ \tx -> do
                         jumanentries <- callJuman' tx
                         S.appendfile filepath $ T.snoc (T.intercalate "," $ map encodeJumanEntry jumanentries) '\n'
                         S.echo_n_err "o" 
                         return $ map jumanEntry2Pair jumanentries 

processTextByJumanOnline :: T.Text -> IO [(Kihon, Hinsi)]
processTextByJumanOnline text =
  S.shelly $ do
             j <- callJuman' text
             return $ map jumanEntry2Pair j

jumanOutput2Entry :: T.Text -> JumanEntry
jumanOutput2Entry tx 
  | tx == "EOS" = EOS
  | otherwise = let ts = T.split (==' ') tx in
                  if length ts < 6
                     then Error tx
                     else JumanEntry {hyoso=ts!!0, yomi=ts!!1, kihon=ts!!2, pos=ts!!3, subpos=ts!!5}

encodeJumanEntry :: JumanEntry -> T.Text
encodeJumanEntry j = case j of
  JumanEntry _ _ _ _ _ -> T.concat [kihon j, "#", pos j]
  EOS                  -> "EOS#EOS"
  Error t              -> T.concat ["error#",t]

jumanEntry2Pair :: JumanEntry -> (Kihon, Hinsi)
jumanEntry2Pair je = case je of
                       JumanEntry _ _ _ _ _ -> (kihon je, pos je)
                       EOS                  -> ("EOS","EOS")
                       Error t              -> (t,"ERROR")

-- | Create a pair (or a tuple) of sorted labels from juman-processed texts
--   ([k1,...,kn],[p1,...,pn])
buildDictionary :: [[(Kihon, Hinsi)]]    -- ^ All the juman output texts: ["k1#p1,...,kn#pn","k1#p1,...,kn#pn",...]
                   -> [Hinsi] -- ^ poss to use.  when posfilter==[], all poss are used.
                   -> ([Kihon],[Hinsi])
buildDictionary allResults posfilter =
  unzip $ do                      -- list monad
          oneResult <- allResults -- [(k1,p1),...,(kn,pn)]
          (k,p)     <- oneResult  -- (k1,p1)
          S.when (posfilter /= []) $ guard $ L.elem p posfilter         -- Only considers elements specified in posfilter
          return (k,p)                           -- [(k1,p1),...,(kn,pn)]

-- | Remove (from a given list) elements whose occurrence is equal or less than the threshold
cutLessFreqWords :: Int      -- ^ The minimum number of occurences of a word (otherwise the word is ignored)
                 -> [T.Text] -- ^ a list of words
                 -> [T.Text]
cutLessFreqWords threshold wds = fst $ unzip $ reverse $ L.sortOn snd $ U.toList $ U.filterHistByValue (>= threshold) $ U.pushWords wds

{-

-- | transforms a list of texts to a list of tensors (to be fed to neural networks)
juman2tensor :: Bool                  -- ^ True for verbose mode, False for silent mode
               -> Int                 -- ^ The minimum number of occurences of a word (otherwise the word is ignored)
               -> ([Kihon],[Hinsi]) -- ^ dictionary (sorted all kihon, sorted all pos)
               -> [[(Kihon, Hinsi)]]            -- ^ A juman output texts to transform into tensors
               -> IO(Int,Int,Int,[[(Int,Int)]])
juman2tensor _ threshold (all_kihon,all_pos) jumanResults = do -- IO monad
  -- filter (\p -> snd p /= "未定義語") 
  let kihon_all_sorted = cutByThreshold all_kihon
      pos_all_sorted   = cutByThreshold all_pos
      jumanEntries = do
                     oneResult <- jumanResults                 -- [(k1,p1),...,(kn,pn)]
                     return $ do
                              (k,p) <- oneResult
                              (i,_) <- U.fetchIndex kihon_all_sorted k
                              (j,_) <- U.fetchIndex pos_all_sorted p
                              return (i,j) -- [(k1,p1),...,(kn,pn)]
  --IO.withFile (datadir </> datafilename ++ "_kihon_all_sorted.txt") IO.WriteMode (\h -> mapM_ (printPair h) kihon_all_sorted)
  --IO.withFile (datadir </> datafilename ++ "_pos_all_sorted.txt") IO.WriteMode (\h -> mapM_ (printPair h) pos_all_sorted)
  return (length kihon_all_sorted,
          length pos_all_sorted,
          maximum $ map length jumanEntries,
          jumanEntries)
  where cutByThreshold wds = reverse $ L.sortOn snd $ U.toList $ U.filterHistByValue (>= threshold) $ U.pushWords wds

unzipJumanTensorWithMask :: FilePath         -- ^ path of directory to save data files
                    -> String        -- ^ filename
                    -> Int
                    -> [[(Int,Int)]] -- ^ A list of (kihon index, pos index)
                    -> IO(String,String)
unzipJumanTensorWithMask datadir datafilename maxlength jumanEntries = do
  let kihonFileName = datafilename ++ "_kihon.txt"
      posFileName = datafilename ++ "_pos.txt" 
  IO.withFile (datadir </> kihonFileName) IO.WriteMode (\hk ->
    IO.withFile (datadir </> posFileName) IO.WriteMode (\hp ->
      mapM_ (\jumanEntry -> do -- [(Int,Int)]
        let kp = unzip jumanEntry -- ([Int],[Int])
            k = fst kp            -- [Int]
            p = snd kp            -- [Int]
        T.hPutStrLn hk $ T.intercalate "," $ map (T.pack . show) k ++ replicate (maxlength-(length k)) "0"
        T.hPutStrLn hp $ T.intercalate "," $ map (T.pack . show) p ++ replicate (maxlength-(length p)) "0"
        ) jumanEntries
      )
    )
  return (kihonFileName, posFileName)

jumanTensor2BagOfWords :: FilePath       -- ^ path of directory to save data files
                          -> String      -- ^ filename
                          -> Int
                          -> Int
                          -> [[(Int,Int)]] -- ^ A list of (kihon index, pos index)
                          -> IO(String)
jumanTensor2BagOfWords datadir datafilename kihonkeiNum hinsiNum jumanEntries = do
  IO.withFile (datadir </> datafilename) IO.WriteMode (\h ->
    mapM_ (\jumanEntry -> do -- [(Int,Int)]
            let kp = unzip jumanEntry -- ([Int],[Int])
                k_hist = U.pushWords $ fst kp
                --p_hist = U.pushWords $ snd kp
                k_bow  = map (\i -> case U.lookup i k_hist of
                                      Just j -> show j
                                      Nothing -> "0") [1..kihonkeiNum]
                --p_bow = map (\i -> case U.lookup i p_hist of
                --                      Just j -> show j
                --                      Nothing -> "0") [1..hinsiNum]
            IO.hPutStrLn h $ L.intercalate "," k_bow
          ) jumanEntries
    )
  return datafilename

{-
printPair :: IO.Handle -> (T.Text,Int) -> IO()
printPair h (text,i) = do
  T.hPutStr h text
  IO.hPutStrLn h $ "," ++ show i
-}

printJumanTensorParams :: IO.Handle -> (Int,Int,Int) -> IO()
printJumanTensorParams h (kihonkei_num, hinsi_num, maxlength) = do
  IO.hPutStr   h "Number of unique kihonkei: "
  IO.hPutStrLn h $ show kihonkei_num
  IO.hPutStr   h "Number of unique hinshi: "
  IO.hPutStrLn h $ show hinsi_num
  IO.hPutStr   h "Maxumum number of words: "
  IO.hPutStrLn h $ show $ maxlength

-}



src/DyNet/Simple/LSTMclassifier.hs
{-# OPTIONS -Wall #-}
{-# LANGUAGE OverloadedStrings, DeriveGeneric #-}

module DyNet.Simple.LSTMclassifier (
  jsemlabels,
  Token,
  Label,
  LSTMparams(..),
  jsemlabels,
  HyperParams(..),
  showHyperParams,
  setLSTMparams,
  buildClassifier,
  buildLSTM
  ) where

import Prelude hiding (words)
import GHC.Generics
import Data.List (intercalate)        --base
import Control.Monad (forM)           --base
import qualified Data.Text as StrictT --text
import qualified Data.Aeson as A      --aeson
import qualified DyNet.Core as D      --dynet-haskell
import qualified DyNet.Expr as D      --dynet-haskell
import qualified DyNet.RNN as D       --dynet-haskell
import qualified DyNet.Dict as D      --dynet-haskell
import qualified DyNet.Train as D     --dynet-haskell
import qualified DyNet.Tools.Juman as Juman --dynet-tools

type Token = StrictT.Text
type Label = StrictT.Text

data LSTMparams = LSTMparams {
  words :: D.Dict Token,
  poss :: D.Dict Token,
  labels :: D.Dict Label,
  wordEmbed :: D.LookupParameter,
  posEmbed :: D.LookupParameter,
  w1 :: D.Parameter,
  w2 :: D.Parameter,
  w3 :: D.Parameter,
  fwdRNN :: D.VanillaLSTMBuilder
  --bwdRNN :: D.VanillaLSTMBuilder
  } 

jsemlabels :: [Label]
jsemlabels = ["YES","NO","UNKNOWN"]

data HyperParams = HyperParams {
  numOfLSTMlayers :: Int, -- ^ layers ToDo: 層数？
  wordEmbedDim :: Int,    -- ^ wordEmbedDim
  posEmbedDim :: Int,     -- ^ posEmbedDim
  lstmDim :: Int,         -- ^ lstmDim
  hidden1Dim :: Int,      -- ^ hidden1Dim
  hidden2Dim :: Int,      -- ^ hidden2Dim
  iteration :: Int,       -- ^ number of iteration
  batchSize :: Int,       -- ^ number of minibatch size
  wordslist :: [Token],   -- ^ words
  posslist :: [Token],    -- ^ part-of-speeches
  labelslist :: [Label]   -- ^ answer labels
  } deriving (Generic, Show)

instance A.ToJSON HyperParams
instance A.FromJSON HyperParams

-- | HyperParamsを文字列化（ハイフンで繋ぐだけ）
showHyperParams :: HyperParams -> String
showHyperParams (HyperParams numOfLSTMlayers wordEmbedDim posEmbedDim lstmDim hidden1Dim hidden2Dim iteration batchSize _ _ _) =
  intercalate "-" $ map show [numOfLSTMlayers,wordEmbedDim,posEmbedDim,lstmDim,hidden1Dim,hidden2Dim,iteration,batchSize]

-- | 
setLSTMparams :: D.Model -> HyperParams -> IO(LSTMparams)
setLSTMparams model (HyperParams numOfLSTMlayers wordEmbedDim posEmbedDim lstmDim hidden1Dim hidden2Dim _ _ words poss labels) =
  LSTMparams <$> D.createDict words (Just "<unkw>") -- ToDo: D.createDictの第二引数は見つからない時に返す値か？
         <*> D.createDict poss (Just "<unkp>")
         <*> D.createDict labels Nothing
         <*> D.addLookupParameters' model (length words + 1) [wordEmbedDim]
         <*> D.addLookupParameters' model (length poss + 1) [posEmbedDim]
         <*> D.addParameters' model [hidden1Dim, lstmDim*3]     -- w1
         <*> D.addParameters' model [hidden2Dim, hidden1Dim]    -- w2
         <*> D.addParameters' model [length labels, hidden2Dim] -- w3
         <*> D.createVanillaLSTMBuilder numOfLSTMlayers (wordEmbedDim+posEmbedDim) lstmDim model False

type JumanWord = (Juman.Kihon,Juman.Hinsi)

buildClassifier :: LSTMparams -> D.ComputationGraph -> ([JumanWord],[JumanWord]) -> IO D.Expression
buildClassifier params cg (premise,hypo) = do
  _W1 <- D.parameter cg (w1 params)
  _W2 <- D.parameter cg (w2 params)
  _W3 <- D.parameter cg (w3 params)
  premiseVec <- last <$> buildLSTM params cg premise
  hypoVec <- last <$> buildLSTM params cg hypo
  mult <- D.cmult premiseVec hypoVec
  _W3 `D.mul` (D.selu (_W2 `D.mul` (D.selu (_W1 `D.mul` (D.concat' [premiseVec, hypoVec, mult])))))

buildLSTM :: LSTMparams -> D.ComputationGraph -> [JumanWord] -> IO [D.Expression]
buildLSTM params cg inputs = do
  D.newGraph' (fwdRNN params) cg -- ToDo: これ何してんだっけ. D.parameter
  wembs <- forM inputs $ \(word,pos) -> do
             wordVec <- D.fromString (words params) word >>= D.lookup cg (wordEmbed params)
             posVec <- D.fromString (poss params) pos >>= D.lookup cg (posEmbed params)
             return $ D.concat' [wordVec, posVec]
  D.startNewSequence' (fwdRNN params) -- ToDo: これは確か初期値ベクトルをセットしている
  mapM (D.addInput (fwdRNN params)) wembs




app/jsem/jsemTrain.hs
{-# LANGUAGE OverloadedStrings, DeriveAnyClass, DeriveGeneric #-}

--module JSeMtrain where

import System.FilePath ((</>),(<.>))     --filepath
import qualified Data.Text as StrictT    --text
import qualified Data.Text.IO as StrictT --text
import qualified Data.Text.Lazy    as T  --text
import qualified Data.Text.Lazy.IO as T  --text
import qualified Data.Aeson as A         --aeson
import qualified DyNet.Core as D         --dynet-haskell
import qualified DyNet.Expr as D         --dynet-haskell
import qualified DyNet.RNN as D          --dynet-haskell
import qualified DyNet.Dict as D         --dynet-haskell
import qualified DyNet.Train as D        --dynet-haskell
import qualified DyNet.Tools.Juman as Juman --dynet-tools
import qualified DyNet.Tools.Chart as D     --dynet-tools
import qualified DyNet.Simple as D          --dynet-tools
import qualified JSeM as J                  --jsem
import qualified JSeM.XML as J              --jsem
import DyNet.Simple.LSTMclassifier          --dynet-tools

main :: IO ()
main = do
  -- | 初期化。引数は捨てる。
  _ <- D.setup
  -- | 変数の宣言。
  let homedir = "/home/bekki/program/dynet-tools"
      jsemfile = "/home/bekki/program/jsem/data/v1.0/GeneralizedQuantifier.xml"
      datadir = homedir </> "app" </> "jsem" </> "jsem_data"
      savedir = homedir </> "app" </> "jsem" </> "jsem_models"
  -- | jsemfileを読み込み、XML形式データをパーズし[JSeM.JSeMdata]に。
  jsemdata <- J.xmlFile2jsemData jsemfile
  --let (ans,premises,hypos) = unzip3 $ map (\j -> (JSeM.answer j, T.concat $ JSeM.premise j, JSeM.hypothesis j)) jsemdata
  let (ans,premises,hypos) = unzip3 $ do
                                      j <- jsemdata
                                      return (J.answer j, StrictT.concat $ J.premises j, J.hypothesis j)
      -- | 解答のリスト:[Label]を作る。
      allAnswers = do
                   a <- ans
                   return $ case a of
                              J.YES -> StrictT.pack "YES"
                              J.NO ->  StrictT.pack "NO"
                              J.UNKNOWN -> StrictT.pack "UNKNOWN"
                              J.UNDEF ->   StrictT.pack "UNKNOWN"
  -- | 入力データをJUMANで形態素解析。各行 -> [(ある,同士),(イタリア,名詞),...]
  allPremises <- Juman.processTextsByJuman (datadir  </> "juman_premises.txt") premises
  allHypos    <- Juman.processTextsByJuman (datadir </> "juman_hypos.txt") hypos 
  -- | 全形態素解析結果から、表層形と品詞のそれぞれについて辞書を作る（頻度がthreshold以下は捨てる）
  let (allwords,allposs) = Juman.buildDictionary (allPremises ++ allHypos) [] -- ([ある,イタリア,...],[動詞,名詞,...])
      threshold = 2
      freqwords = Juman.cutLessFreqWords threshold allwords
      freqposs  = Juman.cutLessFreqWords threshold allposs
  -- | ここまで前処理、ここからDyNet.hsによるニューラルネットの準備
  D.gridsearch $ do -- Grid Search
    -- | ハイパーパラメータ
    numOfLSTMlayers <- [1]
    wordEmbedDim <- [128,256]
    posEmbedDim <- [10]
    lstmDim <- [128,256,524]
    hidden1Dim <- [128,256,410]
    hidden2Dim <- [60,90,120]
    iter <- [50,75,100]
    batchSize <- [264]
    return $ do
      -- | Grid search開始。ハイパーパラメータを設定。
      let hyperParams = HyperParams numOfLSTMlayers wordEmbedDim posEmbedDim lstmDim hidden1Dim hidden2Dim iter batchSize freqwords freqposs jsemlabels
          setting = showHyperParams hyperParams
      -- | ハイパーパラメータをjson形式で保存。
      A.encodeFile (savedir </> setting <.> "json") hyperParams 
      putStrLn $ "Grid: " ++ setting
      -- | モデルを用意。ハイパーパラメータを元にパラメータを設定
      model <- D.createModel
      params <- setLSTMparams model hyperParams
      -- | データをk-foldで訓練データとテストデータに分割
      let (trainData,testData) = D.kfold 3 $ zip (zip allPremises allHypos) allAnswers
      -- | 学習。訓練器を作りtrainに渡す。返り値はloss/epochの表データで、drawLearningCurveでグラフ化してpng保存。
      D.createAdamTrainer' model  -- ToDo: AdamTrainerとは？それ以外の選択肢は？
        >>= D.train iter batchSize trainData (buildClassifier params) (labels params)
        >>= D.drawLearningCurve (savedir </> setting <.> "png") ("Grid-" ++ setting) 
      -- | 評価。返り値はoverall accuracy
      accuracy <- D.evaluate testData (buildClassifier params) (labels params) jsemlabels
      -- | モデルを保存。ハイパーパラメータの値からファイル名を作る
      D.saveModel model $ savedir </> setting <.> "model"
      -- | model:D.Model, setting:String, Accuracy:Float の三つ組を返す。
      return (model, setting, accuracy)

-- | 現在のベストスコア
-- Accuracy: 0.78707224 Setting: 1-128-10-256-410-90-22-264
-- Accuracy: 0.7794677  Setting: 1-128-10-256-412-88-20-100
-- Accuracy: 0.7794677  Setting: 1-128-10-256-410-90-20-100
